Add coronary artery segmentation bundle (CoronSegmentator)

Foxconn-bgroup-aidd opened the pull request:
proposed Model: CoronSegmentator

This PR adds a new coronary artery segmentation model named `coronsegmentator` under `models/coronsegmentator/`.

Coronsegmentator is an automated pipeline that performs dual-task segmentation on cardiac CT images, focusing on both whole-heart and coronary artery structures. It integrates MONAI’s Auto3DSeg for general cardiac segmentation and a custom nnU-Net model for detailed coronary artery segmentation. The pipeline further converts segmentation results (in STL format) into USD files for downstream 3D visualization and digital twin simulation using NVIDIA Omniverse.




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added coronary artery segmentation from CT images using a pretrained neural network model
  * Segmentation results export to 3D formats for digital twin visualization

* **Documentation**
  * Added comprehensive model documentation and metadata specifications
  * Included Apache 2.0 license and third-party attribution

* **Tests**
  * Added unit tests to validate the segmentation pipeline configuration
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Foxconn-bgroup-aidd committed:
Add coronary artery segmentation bundle (CoronSegmentator)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

yiheng-wang-nv commented on models/CoronSegmentator/.gitignore:
```diff
@@ -0,0 +1,7 @@
+Input/
```
Hi @Foxconn-bgroup-aidd , can we remove this file?
A part of files are already ignored (set in https://github.com/Project-MONAI/model-zoo/blob/dev/.gitignore), and for customized folder like Input/ and Output/, I think we can remove them directly. What do you think?

Foxconn-bgroup-aidd replied:
There is no problem removing them. Thanks.

yiheng-wang-nv commented on models/CoronSegmentator/configs/inference.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "bundle_root": "./",
```
could you prepare a unit test script?
It's helpful to check if (1) large files are downloadable (2) inference pipeline works (3) dependencies mentioned in metadata are correct and enough to run the script.

You can refer to this example to prepare the script: https://github.com/Project-MONAI/model-zoo/blob/dev/ci/unit_tests/test_spleen_ct_segmentation.py

tylin7111095022 replied:
@yiheng-wang-nv, I prepare the script refer to above url.  but when I try to run the script , I got the error: `ModuleNotFoundError: Cannot locate class or function path: 'scripts.heart_digital_twin.CoroSegmentator_Pipeline'`. It seems to can't access the custom class is defined in inference.json. I put the script in `ci/unitest/` folder and run the script at model_zoo root directory. How can I solve the problem? Thanks for answer. Following is revised inference.json:

```
{
  "imports": ["$import glob"],
  "bundle_root": "./",
  "ckpt_dir": "$@bundle_root + 'models'",
  "dataset_dir": "",
  "data": "$list(sorted(glob.glob(@dataset_dir + '/*.nii.gz')))",
  "output_ext": ".usd",
  "output_dir": "$@bundle_root + 'Output'",
  "device": "$torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')",
  "inferer": {
    "_target_": "scripts.heart_digital_twin.CoroSegmentator_Pipeline",
    "input_params": {
      "inputFile": "$@data[0]",
      "outputDir": "$@output_dir"
    }
  },

  "dataset": {},
  "evaluator": {},
  "network_def": {},

  "run": ["$@inferer.run()"]
}
```

KumoLiu replied:
Hi @Foxconn-bgroup-aidd, please try create an __init__.py file in the scripts directory to make it as Python packages and also put the scripts folder just under you bundle instead of under the unittest folder. Thanks.

tylin7111095022 replied:
Thanks for comment, @KumoLiu . Following is my directory tree of project

```
model-zoo/
|--ci/
    |--unit_tests/
        |--test_coronaryArtery_ct_segmentation.py
|--models/
    |--myBundle/
        |--large_file.yml
        |--LICENSE
        |--configs/
            |--inference.json
            |--metadata.json
        |--docs/
        |--models/
        |--scripts/
            |--__init__.py
            |--heartDitgitalTwin.py
```

I run the `python ci/unit_tests/test_coronaryArtery_ct_segmentation.py` and got the `ModuleNotFoundError: Cannot locate class or function path: 'scripts.heart_digital_twin.CoroSegmentator_Pipeline'`. I still try to solve the problem which `pydoc.locate(scripts.heart_digital_twin.CoroSegmentator_Pipeline)` can't access custom class. Thanks for help.

tylin7111095022 replied:
@KumoLiu I add the path of bundle into the sys.path and solve it, Thanks!

KumoLiu commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,88 @@
+{
+    "schema": "https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20220324.json",
```
Hi @Foxconn-bgroup-aidd, could you please utilize the latest schema available at the following link:
"https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20240725.json"

Based on the content, it appears that you are already using the most recent version, as it includes "required_packages_version." Thanks.

ericspod replied:
```suggestion
    "schema": "https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20240725.json",
```

KumoLiu commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,88 @@
+    "numpy_version": "2.3.1",
+    "required_packages_version": {
+        "fire": "0.7.0",
+        "monai": " 1.4.0",
+        "nnunetv2": "2.6.0",
+        "nibabel": "5.3.2",
+        "numpy": "2.3.1",
```
Could you please double check the monai version you are using, the one mentioned above is "1.4.0"? Thanks!
```suggestion
        "nnunetv2": "2.6.0",
        "nibabel": "5.3.2",
```

KumoLiu commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,88 @@
+                    1
+                ],
+                "channel_def": {
+                    "0": "Background",
+                    "1": "Coronary Artery"
```
It seems the "num_channels" is only one for the outputs.

KumoLiu commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,88 @@
+            }
+        }
+    },
+    "optional_packages_version": {
+        "monai": "1.3.2",
+        "torch": "2.7.0+cu126",
+        "numpy": "2.3.1"
+    }
```
I guess you can remove this part, the "optional_packages_version" is no longer a required one in the latest schema. thanks.

Foxconn-bgroup-aidd replied:
The version of monai used in our project is “1.4.0”. “optional_packages_version" will be removed.

KumoLiu commented on models/CoronSegmentator/docs/README.md:
````diff
@@ -0,0 +1,63 @@
+python -m monai.bundle run --config_file "configs/inference.json"
+```
+
+The ImageCAS dataset is publicly available at:
+https://github.com/XiaoweiXu/ImageCAS-A-Large-Scale-Dataset-and-Benchmark-for-Coronary-Artery-Segmentation-based-on-CT.git''
````
Would be better to mention the dataset link in the "Data" section, what do you think?

KumoLiu commented on models/CoronSegmentator/requirements.txt:
I don't think we need this requierments file here, all the dependencies should be included in the metadata.json. Thanks.

KumoLiu commented on models/CoronSegmentator/scripts/cardiac_segmentation/cardiac_seg.py:
```diff
@@ -0,0 +1,188 @@
+            tempfile.TemporaryDirectory(prefix=f"CardiacSeg_Input_{self.instance_id}_") as input_tmp_dir,
+            tempfile.TemporaryDirectory(prefix=f"CardiacSeg_Output_{self.instance_id}_") as output_tmp_dir,
+        ):
+            # try:
+            # Step 1: Convert input to NRRD
+            input_nrrd = self._convert_to_nrrd(self.input_path, input_tmp_dir)
+            # Step 2: Run segmentation
+            seg_nrrd = self._run_segmentation(input_nrrd, output_tmp_dir)
+            # Step 3: Process results
+            self._process_segmentation_results(seg_nrrd, output_tmp_dir)
+            # except Exception as e:
+            #     logger.error(f"Processing failed: {str(e)}")
+            #     raise
```
Could you please remove the comment code here or perhaps include the try-catch logic is also make sense.

KumoLiu commented on models/CoronSegmentator/scripts/usd_design/usd_mesh_creator.py:
```diff
@@ -0,0 +1,166 @@
+                    base_center = Gf.Vec3d(
+                        (bbox_min[0] + bbox_max[0]) / 2, bbox_min[1], (bbox_min[2] + bbox_max[2]) / 2
+                    )
+                    # 计算平移量，使得基底中心移动到原点
```
Please check and remove all Chinese comments inline. Thanks.

KumoLiu commented on models/CoronSegmentator/configs/inference.json:
```diff
@@ -0,0 +1,17 @@
+    "output_ext": ".usd",
+    "input_file": "Input/E_CT_0000.nii.gz",
+    "infer": {
+        "_target_": "scripts.heart_digital_twin.CoroSegmentator_Pipeline",
```
As @yiheng-wang-nv mentioned above, please include the unit test to ensure the bundle works as expected.
I didn't see this file under the scripts folder.

KumoLiu reviewed (commented):
Hi @Foxconn-bgroup-aidd, thanks for the contributions! Leave several comments inline.

Foxconn-bgroup-aidd committed:
1. Add unit_test scripts 2. Add download.py 3. Update large_file.yml and metadata.json

Foxconn-bgroup-aidd committed:
Merge branch '760-CoronSegmentator' of github.com:Foxconn-bgroup-aidd/model-zoo into 760-CoronSegmentator

Foxconn-bgroup-aidd committed:
Merge branch '760-CoronSegmentator' of github.com:Foxconn-bgroup-aidd/model-zoo into 760-CoronSegmentator

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

KumoLiu commented on ci/unit_tests/test_brain_image_synthesis_latent_diffusion_model.py:
Did you mis-delete this test file?

Foxconn-bgroup-aidd replied:
We have fixed the issue of incorrectly deleting test files.

Foxconn-bgroup-aidd replied:
Hi, I’ve pushed the updated code. Thank you for your feedback so far! Please let me know if there’s anything else I should work on.

Foxconn-bgroup-aidd committed:
Restore deleted unit test: test_brain_image_synthesis_latent_diffusion_model.py

Foxconn-bgroup-aidd committed:
Restore deleted test_brain_image_synthesis_latent_diffusion_model.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

yiheng-wang-nv committed:
Merge branch 'dev' into 760-CoronSegmentator

yiheng-wang-nv commented:
HI @Foxconn-bgroup-aidd , based on the pre-commit CI, it seems the following json files have format issues:
```
Input File models/CoronSegmentator/configs/metadata.json is not a valid JSON, consider using check-json
Input File models/CoronSegmentator/configs/inference.json is not a valid JSON, consider using check-json

```

Foxconn-bgroup-aidd committed:
Update metadata.json

Signed-off-by: Foxconn-bgroup-aidd <bgroup-aidd@foxconn.com>

Foxconn-bgroup-aidd committed:
Update inference.json

Signed-off-by: Foxconn-bgroup-aidd <bgroup-aidd@foxconn.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Foxconn-bgroup-aidd committed:
Update metadata.json

Signed-off-by: Foxconn-bgroup-aidd <bgroup-aidd@foxconn.com>

Foxconn-bgroup-aidd committed:
Update metadata.json

Signed-off-by: Foxconn-bgroup-aidd <bgroup-aidd@foxconn.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

yiheng-wang-nv committed:
Merge branch 'dev' into 760-CoronSegmentator

Foxconn-bgroup-aidd committed:
Apply isort import fixes

ericspod commented:
Hi @Foxconn-bgroup-aidd we still want to accept this model so please have a look at outstanding conversations and resolve those that have been addressed. @KumoLiu and @yiheng-wang-nv can continue their review from there. Thanks!

Foxconn-bgroup-aidd committed:
1.Revise the pytorch version in configs/metadata.json 2.Pass the isort,  black, flake8 and pytype codeformat test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Foxconn-bgroup-aidd committed:
1.Revise the package version in configs/metadata.json 2.Pass the flake8 test

yiheng-wang-nv committed:
Merge branch 'dev' into 760-CoronSegmentator

yiheng-wang-nv commented:
/build

yiheng-wang-nv commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,81 @@
+        "monai": " 1.5.0",
+        "nnunetv2": "2.6.2",
+        "nibabel": "5.3.2",
+        "numpy": "2.3.3",
```
Hi @Foxconn-bgroup-aidd , can you confirm the numpy version here? I saw CI error: https://github.com/Project-MONAI/model-zoo/actions/runs/18928765961/job/54041291172?pr=761#step:7:488 It may due to wrong numpy version, or the python version used in CI does not support this version. May I know which python version do you use?

Foxconn-bgroup-aidd replied:
Hi, we are using Python 3.12.9, and we have already pushed a new version. Please let us know if you need any additional information

yiheng-wang-nv replied:
Hi @ericspod , can you help to check this PR? We may need to enable tests on new python versions. I'm very busy on other projects, and may not have enough time for model zoo recently.

Foxconn-bgroup-aidd committed:
Update metadata version

ericspod committed:
Merge branch 'dev' into 760-CoronSegmentator

ericspod commented on models/CoronSegmentator/large_file.yml:
```diff
@@ -0,0 +1,11 @@
+large_files:
+  - path: "models/cardiacModel.pt"
+    # url: "https://mega.nz/file/L8oz3QhA#illHJWSCmsmV5xqbXktwiu-1jUebbpJ0HSKv0hbMFCc"
```
Please remove these commented links.

ericspod commented on models/CoronSegmentator/scripts/download.py:
```diff
@@ -0,0 +1,59 @@
+    return md5.hexdigest()
+
+
+def _download(url, destination_path):
```
MONAI has it's own [download routines](https://github.com/Project-MONAI/MONAI/blob/c1d02cf1e0bf1e9e7a62ead7816c2aff30d93be6/monai/apps/utils.py#L191) which you can use instead, and which also do hash checking.

ericspod reviewed (commented):
I had just very minor comments to make. @KumoLiu had mentioned a few things previously, if these have been addressed we can marked those comments as resolved. Once the tests pass we can add this.

ericspod commented:
@KumoLiu we seem to be having an issue with space for the tests, did we have a solution for this in this repo? Thanks!

peter-ty-lin committed:
Only preserve the coronary artery segmentation module to solve the problem which  remaining space is not enough.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ericspod committed:
Merge branch 'dev' into 760-CoronSegmentator

ericspod commented:
Hi @Foxconn-bgroup-aidd sorry for the long delay on this one, please have a look at the comments made by us and Coderabbit and mark resolved anything that's sorted. I think there's some valid concerns about the tests that should be addressed. Thanks!

Foxconn-bgroup-aidd committed:
fix: address review comments for CoronSegmentator bundle

- Replace custom download.py with monai.apps.utils.download_url
- Add __init__.py to all script sub-packages for proper Python packaging
- Fix code style: black, isort, flake8 (max-line 120)
- Remove commented-out dead code (logger.error)
- Fix metadata.json: remove leading space in monai version
- Fix docs/README.md: remove trailing quotes
- Fix unit test: cross-platform path, int16 dtype, run=[] CI override

Foxconn-bgroup-aidd commented on models/CoronSegmentator/scripts/usd_design/usd_mesh_creator.py:
```diff
@@ -0,0 +1,179 @@
+            for input_file, mesh in zip(self.input_files, meshes):
+                xform_path, mesh_path = self.get_paths(input_file)
+                if self.create_usd_mesh(mesh, xform_path, mesh_path):
+                    mesh_paths.append(mesh_path)
```
This file has been removed in the latest commit as part of the pipeline simplification. No longer applicable.

Foxconn-bgroup-aidd committed:
fix: address CodeRabbit review — quote variables, fix is_pip_installed, add --version flag

Foxconn-bgroup-aidd commented:
Thanks for the detailed feedback!
All comments have been addressed in this update, and the CodeRabbitAI review has passed.

Please let us know if there’s anything else to fix.

ericspod commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,78 @@
+        "inputs": {
+            "image": {
+                "type": "image",
+                "format": "hounsfield",
+                "modality": "CT",
+                "num_channels": 1,
+                "spatial_shape": [
+                    512,
+                    512,
+                    256
+                ],
+                "dtype": "float32",
+                "value_range": [
+                    0,
+                    1
+                ],
```
The images don't appear to be scaled in your scripts so the assumption is they pixel data in the Hounsfield range as described here.

ericspod commented on models/CoronSegmentator/docs/README.md:
````diff
@@ -0,0 +1,63 @@
+Input: .nii.gz NIfTI format (CT scan)
+
+Resolution normalization handled internally
+
+No manual annotation required for inference
+
+#### **Inference**
+
+```bash
+python -m monai.bundle run --config_file "configs/inference.json"
+```
````
This should also be changed to include these requirements in the command line.

ericspod commented on models/CoronSegmentator/runtests.sh:
Is this the same file as the one in the root of the repository? If so this shouldn't be here, if it's different it should go into the `docs` directory.

ericspod commented on models/CoronSegmentator/large_files.yml:
```diff
@@ -0,0 +1,11 @@
+large_files:
+  - path: "models/cardiacModel.pt"
+    # url: "https://mega.nz/file/L8oz3QhA#illHJWSCmsmV5xqbXktwiu-1jUebbpJ0HSKv0hbMFCc"
+    url: "https://drive.google.com/file/d/11-pgULhmSZyAAfZG9n1vOusWHeH-hpwE/view?usp=drive_link"
```
We would much prefer weights to be hosted elsewhere such as Huggingface.

ericspod commented on models/CoronSegmentator/large_files.yml:
```diff
@@ -0,0 +1,11 @@
+    # url: "https://mega.nz/file/L8oz3QhA#illHJWSCmsmV5xqbXktwiu-1jUebbpJ0HSKv0hbMFCc"
+    url: "https://drive.google.com/file/d/11-pgULhmSZyAAfZG9n1vOusWHeH-hpwE/view?usp=drive_link"
+    hash_val: "0e783497e94293c86995392e5fbbc790"
+    hash_type: "md5"
```
I would suggest checking the has would be a good idea as described here.

ericspod commented on models/CoronSegmentator/large_file.yml:
```diff
@@ -0,0 +1,9 @@
+large_files:
+  - path: "models/cardiacModel.pt"
```
```suggestion
  - path: "models/model.pt"
```
This is the expected model file name the testing system looks for.

ericspod committed:
Merge branch 'dev' into 760-CoronSegmentator

ericspod commented:
Hi @Foxconn-bgroup-aidd please have a look at the comments above from Coderabbit regarding some minor changes, and the errors from testing which should be very quick fixes. These looks like easy fixes so we can get this merged soon hopefully.

Foxconn-bgroup-aidd committed:
fix: apply black formatting, fix flake8 E203, address CodeRabbit review comments

Foxconn-bgroup-aidd committed:
fix: resolve CI failures for CoronSegmentator bundle

- rename large_file.yml to large_files.yml so CI can locate and download large files
- add output file existence check before shutil.move to surface missing prediction errors clearly
- reformat scripts with black --skip-magic-trailing-comma to match CI format check

ericspod committed:
Merge branch 'dev' into 760-CoronSegmentator

ericspod commented on ci/unit_tests/test_coronaryArtery_ct_segmentation.py:
```diff
@@ -0,0 +1,89 @@
+#!/bin/python
```
```suggestion
```

ericspod committed:
Apply suggestion from @ericspod

Signed-off-by: Eric Kerfoot <17726042+ericspod@users.noreply.github.com>

ericspod commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,80 @@
+                "dtype": "float32",
+                "value_range": [
+                    0,
+                    1
```
```suggestion
                    -1024,
                    3071
```

ericspod committed:
Merge branch 'dev' into 760-CoronSegmentator

ericspod commented on models/CoronSegmentator/configs/metadata.json:
```diff
@@ -0,0 +1,80 @@
+    "required_packages_version": {
+        "fire": "0.7.0",
+        "matplotlib": "3.10.7",
+        "monai": "1.5.0",
```
```suggestion
    "monai_version": "1.6.0",
    "pytorch_version": "2.6.0",
    "numpy_version": "2.2.6",
    "required_packages_version": {
        "fire": "0.7.0",
        "matplotlib": "3.10.7",
```
